File: src/main/scala/com/johnsnowlabs/nlp/annotators/er/EntityRuler.scala
/*
 * Copyright 2017-2021 John Snow Labs
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.johnsnowlabs.nlp.annotators.er

import com.johnsnowlabs.nlp.AnnotatorApproach
import com.johnsnowlabs.nlp.AnnotatorType.{CHUNK, DOCUMENT, TOKEN}
import com.johnsnowlabs.nlp.annotators.param.ExternalResourceParam
import com.johnsnowlabs.nlp.util.io.ResourceHelper.spark
import com.johnsnowlabs.nlp.util.io.{ExternalResource, ReadAs, ResourceHelper}
import com.johnsnowlabs.storage.Database.Name
import com.johnsnowlabs.storage.{Database, HasStorage, RocksDBConnection, StorageReadWriter, StorageWriter}
import com.johnsnowlabs.util.JsonParser
import org.apache.spark.ml.PipelineModel
import org.apache.spark.ml.param.BooleanParam
import org.apache.spark.ml.util.Identifiable
import org.apache.spark.sql.{DataFrame, Dataset}
import org.apache.spark.sql.functions.collect_set

import scala.io.Source

class EntityRuler(override val uid: String) extends AnnotatorApproach[EntityRulerModel] with HasStorage {

  def this() = this(Identifiable.randomUID("ENTITY_RULER"))

  override val description: String = "Entity Ruler matches entities based on text patterns"

  val patternsResource = new ExternalResourceParam(this, "patternsResource",
    "Resource in JSON or CSV format to map entities to patterns")

  val enablePatternRegex = new BooleanParam(this, "enablePatternRegex",
    "Enables regex pattern match")

  def setEnablePatternRegex(value: Boolean) : this.type = set(enablePatternRegex, value)

  private var entities: Array[String] = Array()

  def setPatternsResource(path: String, readAs: ReadAs.Format,
                          options: Map[String, String] = Map("format" -> "JSON")): this.type =
    set(patternsResource, ExternalResource(path, readAs, options))

  setDefault(storagePath -> ExternalResource("", ReadAs.TEXT, Map()),
    patternsResource -> null, enablePatternRegex -> false
  )

  private val AVAILABLE_FORMATS = Array("JSON", "JSONL", "CSV")

  override def train(dataset: Dataset[_], recursivePipeline: Option[PipelineModel]): EntityRulerModel = {

    val entityRuler = new EntityRulerModel()
      .setStorageRef($(storageRef))
      .setEnablePatternRegex($(enablePatternRegex))

    if ($(enablePatternRegex)) {
      entityRuler.setRegexEntities(entities)
    }
    entityRuler
  }

  /** Annotator reference id. Used to identify elements in metadata or to refer to this annotator type */
  override val inputAnnotatorTypes: Array[String] = Array(DOCUMENT, TOKEN)
  override val outputAnnotatorType: AnnotatorType = CHUNK
  override protected val databases: Array[Name] = EntityRulerModel.databases

  protected def index(fitDataset: Dataset[_], storageSourcePath: Option[String], readAs: Option[ReadAs.Value],
                               writers: Map[Name, StorageWriter[_]], readOptions: Option[Map[String, String]]): Unit = {

    require($(patternsResource) != null, "patternsResource parameter required")
    require($(patternsResource).path != "", "path for a patternsResource file is required")
    require(AVAILABLE_FORMATS.contains($(patternsResource).options.getOrElse("format", "").toUpperCase()),
      "format option parameter required with either JSON or CSV values")
    if ($(patternsResource).options("format").toUpperCase() == "CSV") {
      require($(patternsResource).options.getOrElse("delimiter", "") != "", "delimiter option parameter required")
    }
    require($(patternsResource).readAs != null, "readAs parameter required")

    var storageWriter: StorageReadWriter[_] = null

    if ($(enablePatternRegex)) {
      storageWriter = writers(Database.ENTITY_REGEX_PATTERNS).asInstanceOf[RegexPatternsReadWriter]
    } else {
      storageWriter = writers(Database.ENTITY_PATTERNS).asInstanceOf[PatternsReadWriter]
    }

    resourceFormats match {
      case "JSON&TEXT" => storePatternsFromJson(storageWriter)
      case "JSONL&TEXT" => storePatternsFromJsonl(storageWriter)
      case "JSON&SPARK" => storePatternsFromJSONDataFrame(storageWriter, "JSON")
      case "JSONL&SPARK" => storePatternsFromJSONDataFrame(storageWriter, "JSONL")
      case "CSV&TEXT" => storePatternsFromCSV(storageWriter)
      case "CSV&SPARK" => storeEntityPatternsFromCSVDataFrame(storageWriter)
      case _ @ format => throw new IllegalArgumentException(s"format $format not available")
    }

  }

  private lazy val resourceFormats: String = $(patternsResource).options("format").toUpperCase() + "&" +  $(patternsResource).readAs

  private def storePatternsFromJson(storageReadWriter: StorageReadWriter[_]): Unit = {
    import io.circe.generic.auto._

    val stream =  ResourceHelper.getResourceStream($(patternsResource).path)
    val jsonContent = Source.fromInputStream(stream).mkString
    val jsonParser = new JsonParser[EntityPattern]
    val entityPatterns: Array[EntityPattern] = jsonParser.readJsonArray(jsonContent)

    entityPatterns.foreach{ entityPattern =>
      val entity = if (entityPattern.id.isDefined) s"${entityPattern.label},${entityPattern.id.get}" else entityPattern.label
      storageReadWriter match {
        case patternsWriter: PatternsReadWriter => storePatterns(entityPattern.patterns.toIterator, entity, patternsWriter)
        case regexPatternsWriter: RegexPatternsReadWriter => storeRegexPattern(entityPattern.patterns, entity, regexPatternsWriter)
      }
    }
  }

  private def storePatternsFromJsonl(storageReadWriter: StorageReadWriter[_]): Unit = {

    import io.circe.generic.auto._
    val jsonParser = new JsonParser[EntityPattern]

    val sourceStream = ResourceHelper.SourceStream($(patternsResource).path)
    sourceStream.content.foreach( content => content.foreach{ line =>
      val entityPattern: EntityPattern = jsonParser.readJson(line)
      val entity = if (entityPattern.id.isDefined) s"${entityPattern.label},${entityPattern.id.get}" else entityPattern.label
      storageReadWriter match {
        case patternsWriter: PatternsReadWriter => storePatterns(entityPattern.patterns.toIterator, entity, patternsWriter)
        case regexPatternsWriter: RegexPatternsReadWriter => storeRegexPattern(entityPattern.patterns, entity, regexPatternsWriter)
      }
    })
  }

  private def storePatternsFromCSV(storageReadWriter: StorageReadWriter[_]): Unit = {
    storageReadWriter match {
      case patternsWriter: PatternsReadWriter =>
        val entityPatterns: Map[String, String] = ResourceHelper.parseKeyValueText($(patternsResource))
        entityPatterns.foreach(entityPattern => storePattern(entityPattern._2, entityPattern._1, patternsWriter))
      case regexPatternsWriter: RegexPatternsReadWriter =>
        val entityPatterns: Map[String, List[String]] = ResourceHelper.parseKeyListValues($(patternsResource))
        entityPatterns.foreach(entityPattern => storeRegexPattern(entityPattern._2, entityPattern._1, regexPatternsWriter))
    }
  }

  private def storeEntityPatternsFromCSVDataFrame(storageReadWriter: StorageReadWriter[_]): Unit = {

    val patternOptions = $(patternsResource).options
    var patternsDataFrame = spark.read.options(patternOptions)
      .format(patternOptions("format"))
      .options(patternOptions)
      .option("delimiter", patternOptions("delimiter"))
      .load($(patternsResource).path)
      .toDF("label", "pattern")

    patternsDataFrame = patternsDataFrame.groupBy("label")
      .agg(collect_set("pattern").alias("patterns"))

    storeFromDataFrame(patternsDataFrame, storageReadWriter)

  }

  private def storePatternsFromJSONDataFrame(storageReadWriter: StorageReadWriter[_], format: String): Unit = {

    val path = $(patternsResource).path
    var patternsDataFrame = spark.read.option("multiline", "true").json(path)

    if (format.equals("JSONL")) {
      patternsDataFrame = spark.read.json(path)
    }

    storeFromDataFrame(patternsDataFrame, storageReadWriter)
  }

  private def storeFromDataFrame(patternsDataFrame: DataFrame, storageReadWriter: StorageReadWriter[_]): Unit = {
    val fieldId = patternsDataFrame.schema.fields.filter(field => field.name == "id")
    var id = ""

    patternsDataFrame.rdd.toLocalIterator.foreach{ row =>
      val patterns = row.getAs[Seq[String]]("patterns")
      val label = row.getAs[String]("label")
      if (fieldId.nonEmpty) {
        id = row.getAs[String]("id")
      }
      val entity = if (fieldId.nonEmpty) s"$label,$id" else label
      storageReadWriter match {
        case patternsWriter: PatternsReadWriter => storePatterns(patterns.toIterator, entity, patternsWriter)
        case regexPatternsWriter: RegexPatternsReadWriter => storeRegexPattern(patterns, entity, regexPatternsWriter)
      }
    }
  }

  private def storePatterns(patterns: Iterator[String], entity: String, patternsReaderWriter: PatternsReadWriter): Unit = {
    patterns.foreach(pattern => storePattern(pattern, entity, patternsReaderWriter))
  }

  private def storePattern(pattern: String, entity: String, patternsReaderWriter: PatternsReadWriter): Unit = {
    patternsReaderWriter.lookup(pattern).getOrElse(patternsReaderWriter.add(pattern, entity))
  }

  private def storeRegexPattern(pattern: Seq[String], entity: String,
                                regexPatternsReaderWriter: RegexPatternsReadWriter): Unit = {

    if (!entities.contains(entity)) {
      entities = entities ++ Array(entity)
    }
    regexPatternsReaderWriter.lookup(entity).getOrElse(regexPatternsReaderWriter.add(entity, pattern))
  }

  protected def createWriter(database: Name, connection: RocksDBConnection): StorageWriter[_] = {
    database match {
      case Database.ENTITY_PATTERNS => new PatternsReadWriter(connection)
      case Database.ENTITY_REGEX_PATTERNS => new RegexPatternsReadWriter(connection)
    }
  }

}

File: python/sparknlp/annotator/er/entity_ruler.py
#  Copyright 2017-2022 John Snow Labs
#
#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.
"""Contains classes for the EntityRuler."""

from sparknlp.common import *


class EntityRulerApproach(AnnotatorApproach, HasStorage):
    """Fits an Annotator to match exact strings or regex patterns provided in a
    file against a Document and assigns them an named entity. The definitions
    can contain any number of named entities.

    There are multiple ways and formats to set the extraction resource. It is
    possible to set it either as a "JSON", "JSONL" or "CSV" file. A path to the
    file needs to be provided to ``setPatternsResource``. The file format needs
    to be set as the "format" field in the ``option`` parameter map and
    depending on the file type, additional parameters might need to be set.

    If the file is in a JSON format, then the rule definitions need to be given
    in a list with the fields "id", "label" and "patterns"::

         [
            {
              "id": "person-regex",
              "label": "PERSON",
              "patterns": ["\\w+\\s\\w+", "\\w+-\\w+"]
            },
            {
              "id": "locations-words",
              "label": "LOCATION",
              "patterns": ["Winterfell"]
            }
        ]

    The same fields also apply to a file in the JSONL format::

        {"id": "names-with-j", "label": "PERSON", "patterns": ["Jon", "John", "John Snow"]}
        {"id": "names-with-s", "label": "PERSON", "patterns": ["Stark", "Snow"]}
        {"id": "names-with-e", "label": "PERSON", "patterns": ["Eddard", "Eddard Stark"]}

    In order to use a CSV file, an additional parameter "delimiter" needs to be
    set. In this case, the delimiter might be set by using
    ``.setPatternsResource("patterns.csv", ReadAs.TEXT, {"format": "csv", "delimiter": "|")})``::

        PERSON|Jon
        PERSON|John
        PERSON|John Snow
        LOCATION|Winterfell

    ====================== ======================
    Input Annotation types Output Annotation type
    ====================== ======================
    ``DOCUMENT, TOKEN``    ``CHUNK``
    ====================== ======================

    Parameters
    ----------
    patternsResource
        Resource in JSON or CSV format to map entities to patterns
    useStorage
        Whether to use RocksDB storage to serialize patterns

    Examples
    --------
    >>> import sparknlp
    >>> from sparknlp.base import *
    >>> from sparknlp.annotator import *
    >>> from sparknlp.common import *
    >>> from pyspark.ml import Pipeline

    In this example, the entities file as the form of::

        PERSON|Jon
        PERSON|John
        PERSON|John Snow
        LOCATION|Winterfell

    where each line represents an entity and the associated string delimited by "|".

    >>> documentAssembler = DocumentAssembler() \\
    ...     .setInputCol("text") \\
    ...     .setOutputCol("document")
    >>> tokenizer = Tokenizer() \\
    ...     .setInputCols(["document"]) \\
    ...     .setOutputCol("token")
    >>> entityRuler = EntityRulerApproach() \\
    ...     .setInputCols(["document", "token"]) \\
    ...     .setOutputCol("entities") \\
    ...     .setPatternsResource(
    ...       "patterns.csv",
    ...       ReadAs.TEXT,
    ...       {"format": "csv", "delimiter": "\\\\|"}
    ...     )
    >>> pipeline = Pipeline().setStages([
    ...     documentAssembler,
    ...     tokenizer,
    ...     entityRuler
    ... ])
    >>> data = spark.createDataFrame([["Jon Snow wants to be lord of Winterfell."]]).toDF("text")
    >>> result = pipeline.fit(data).transform(data)
    >>> result.selectExpr("explode(entities)").show(truncate=False)
    +--------------------------------------------------------------------+
    |col                                                                 |
    +--------------------------------------------------------------------+
    |[chunk, 0, 2, Jon, [entity -> PERSON, sentence -> 0], []]           |
    |[chunk, 29, 38, Winterfell, [entity -> LOCATION, sentence -> 0], []]|
    +--------------------------------------------------------------------+
    """
    name = "EntityRulerApproach"

    inputAnnotatorTypes = [AnnotatorType.DOCUMENT]
    optionalInputAnnotatorTypes = [AnnotatorType.TOKEN]

    outputAnnotatorType = AnnotatorType.CHUNK

    patternsResource = Param(Params._dummy(),
                             "patternsResource",
                             "Resource in JSON or CSV format to map entities to patterns",
                             typeConverter=TypeConverters.identity)

    useStorage = Param(Params._dummy(),
                       "useStorage",
                       "Whether to use RocksDB storage to serialize patterns",
                       typeConverter=TypeConverters.toBoolean)

    sentenceMatch = Param(Params._dummy(),
                          "sentenceMatch",
                          "Whether to find match at sentence level. True: sentence level. False: token level",
                          typeConverter=TypeConverters.toBoolean)

    alphabet = Param(Params._dummy(),
                     "alphabet",
                     "Alphabet resource path to plain text file with all characters in a given alphabet",
                     typeConverter=TypeConverters.identity)

    @keyword_only
    def __init__(self):
        super(EntityRulerApproach, self).__init__(
            classname="com.johnsnowlabs.nlp.annotators.er.EntityRulerApproach")

    def setPatternsResource(self, path, read_as=ReadAs.TEXT, options={"format": "JSON"}):
        """Sets Resource in JSON or CSV format to map entities to patterns.

        Parameters
        ----------
        path : str
            Path to the resource
        read_as : str, optional
            How to interpret the resource, by default ReadAs.TEXT
        options : dict, optional
            Options for parsing the resource, by default {"format": "JSON"}
        """
        return self._set(patternsResource=ExternalResource(path, read_as, options))

    def setUseStorage(self, value):
        """Sets whether to use RocksDB storage to serialize patterns.

        Parameters
        ----------
        value : bool
            Whether to use RocksDB storage to serialize patterns.
        """
        return self._set(useStorage=value)

    def setSentenceMatch(self, value):
        """Sets whether to find match at sentence level.

        Parameters
        ----------
        value : bool
            True: sentence level. False: token level
        """
        return self._set(sentenceMatch=value)

    def setAlphabetResource(self, path):
        """Alphabet Resource (a simple plain text with all language characters)

        Parameters
        ----------
        path : str
            Path to the resource
        """
        return self._set(alphabet=ExternalResource(path, read_as=ReadAs.TEXT, options={}))

    def _create_model(self, java_model):
        return EntityRulerModel(java_model=java_model)


class EntityRulerModel(AnnotatorModel, HasStorageModel):
    """Instantiated model of the EntityRulerApproach.
    For usage and examples see the documentation of the main class.

    ====================== ======================
    Input Annotation types Output Annotation type
    ====================== ======================
    ``DOCUMENT, TOKEN``    ``CHUNK``
    ====================== ======================
    """
    name = "EntityRulerModel"
    database = ['ENTITY_PATTERNS']

    inputAnnotatorTypes = [AnnotatorType.DOCUMENT]
    optionalInputAnnotatorTypes = [AnnotatorType.TOKEN]

    outputAnnotatorType = AnnotatorType.CHUNK

    def __init__(self, classname="com.johnsnowlabs.nlp.annotators.er.EntityRulerModel", java_model=None):
        super(EntityRulerModel, self).__init__(
            classname=classname,
            java_model=java_model
        )

    @staticmethod
    def pretrained(name, lang="en", remote_loc=None):
        from sparknlp.pretrained import ResourceDownloader
        return ResourceDownloader.downloadModel(EntityRulerModel, name, lang, remote_loc)

    @staticmethod
    def loadStorage(path, spark, storage_ref):
        HasStorageModel.loadStorages(path, spark, storage_ref, EntityRulerModel.database)

